File: packages/engine/src/perception/ctl_optimizer.ts
// CTL Perception Optimizer — implements the OptimizerLoop from aura-x-engine.
//
// Given a CTL, iteratively adjusts instrumentation weights, groove pattern
// transient density, and log drum density curves until the CTL predicts a
// harmonic perception state (C1 + C2 + C3 all satisfied).
//
// This runs BEFORE audio generation — no audio rendering required.
// The agent pipeline becomes:
//
//   synthesizeCtlFromGoal → optimizeCTLForHarmonicState → applyHarmonyPlan → generate

import type { CTLv1 } from "@aura-x/ctl";
import { clamp } from "../_utils";
import {
  predictCTLPerceptionState,
  extractCTLPerceptionParams,
  C1_B_EFF_HARMONIC_MAX,
  C1_OPTIMAL_ALPHA_B,
  C2_MAX_TRANSIENTS,
  C3_MIN_LD_DENSITY,
} from "./ctl_perception_bridge";
import type { CTLPerceptionReport } from "./ctl_perception_bridge";

// ── Types ─────────────────────────────────────────────────────────────────────

export interface PerceptionOptimizationResult {
  ctl:               CTLv1;
  converged:         boolean;
  iterations:        number;
  initial_state:     CTLPerceptionReport;
  final_state:       CTLPerceptionReport;
  mutations_applied: string[];
}

// ── C1 fix: reduce bass dominance ─────────────────────────────────────────────
// Scale down log_drum body_weight so that alpha_B × gain_ld < C1_B_EFF_HARMONIC_MAX.
// If alpha_B is the problem, also soften kick and bass weights.

const C1_LD_SCALE  = 0.88;   // per-iteration scaling factor for log drum
const C1_BASS_SCALE = 0.85;  // per-iteration scaling factor for other bass

function fixC1(ctl: CTLv1, params: ReturnType<typeof extractCTLPerceptionParams>): {
  ctl: CTLv1; mutation: string;
} {
  // Target: alpha_B ≤ C1_OPTIMAL_ALPHA_B (0.32) so b_eff lands safely below 0.40
  const instr = ctl.instrumentation.map(inst => {
    if (inst.family === "log_drum") {
      return { ...inst, body_weight: clamp(inst.body_weight * C1_LD_SCALE) };
    }
    if ((inst.family === "bass" || inst.family === "kick") && params.alpha_B > 0.40) {
      return { ...inst, body_weight: clamp(inst.body_weight * C1_BASS_SCALE) };
    }
    return inst;
  });

  const newB = extractCTLPerceptionParams({ ...ctl, instrumentation: instr }).b_eff;
  return {
    ctl: { ...ctl, instrumentation: instr },
    mutation: `c1_fix: scaled bass family weights → B_eff ${params.b_eff.toFixed(3)}→${newB.toFixed(3)}`,
  };
}

// ── C2 fix: prune transients ──────────────────────────────────────────────────
// Remove non-primary K+L tokens until ≤ 4 per bar.
// Primary positions (0-indexed 16-step bar): 0 (beat 1), 8 (beat 3).
// Pass 1: removes K at non-primary positions.
// Pass 2: removes L at non-primary positions (handles log-drum-dense patterns
//         like three_step where all non-primary K tokens are already gone).

const PRIMARY_TRANSIENT_POSITIONS = new Set([0, 8]);

function fixC2(ctl: CTLv1, params: ReturnType<typeof extractCTLPerceptionParams>): {
  ctl: CTLv1; mutation: string;
} {
  const before = params.transient_density;

  const patterns = ctl.groove_patterns.map(p => {
    const steps = Array.from(p.steps as readonly string[]);
    let count = steps.filter(s => s === "K" || s === "L").length;

    // Pass 1: remove non-primary kicks, back to front
    for (let i = steps.length - 1; i >= 0 && count > C2_MAX_TRANSIENTS; i--) {
      if (steps[i] === "K" && !PRIMARY_TRANSIENT_POSITIONS.has(i)) {
        steps[i] = "-";
        count--;
      }
    }

    // Pass 2: if still over threshold, remove non-primary log drum hits
    for (let i = steps.length - 1; i >= 0 && count > C2_MAX_TRANSIENTS; i--) {
      if (steps[i] === "L" && !PRIMARY_TRANSIENT_POSITIONS.has(i)) {
        steps[i] = "g";  // downgrade to ghost hit — preserves some anchor presence
        count--;
      }
    }

    return { ...p, steps: steps as CTLv1["groove_patterns"][0]["steps"] };
  });

  const after = patterns.length > 0
    ? patterns.reduce((s, p) => s + (p.steps as readonly string[]).filter(t => t === "K" || t === "L").length, 0) / patterns.length
    : 0;

  return {
    ctl: { ...ctl, groove_patterns: patterns },
    mutation: `c2_fix: pruned transients → ${before.toFixed(1)}→${after.toFixed(1)}/bar`,
  };
}

// ── C3 fix: restore anchor continuity ────────────────────────────────────────
// Raise any log_drum_density curve point below the floor.

function fixC3(ctl: CTLv1): { ctl: CTLv1; mutation: string } {
  const ldCurve = ctl.curves.log_drum_density.map(pt =>
    pt.value < C3_MIN_LD_DENSITY
      ? { ...pt, value: C3_MIN_LD_DENSITY }
      : pt
  );
  const changed = ldCurve.filter(
    (pt, i) => pt.value !== ctl.curves.log_drum_density[i].value
  ).length;

  return {
    ctl: { ...ctl, curves: { ...ctl.curves, log_drum_density: ldCurve } },
    mutation: `c3_fix: raised ${changed} log drum density point${changed !== 1 ? "s" : ""} to floor ${C3_MIN_LD_DENSITY}`,
  };
}

// ── Public optimizer ──────────────────────────────────────────────────────────

const MAX_ITERATIONS = 6;

export function optimizeCTLForHarmonicState(
  ctl: CTLv1
): PerceptionOptimizationResult {
  let current       = ctl;
  const mutations: string[] = [];
  const initial_state = predictCTLPerceptionState(current);

  // Already converged — return unchanged
  if (initial_state.state === "harmonic") {
    return {
      ctl: current,
      converged: true,
      iterations: 0,
      initial_state,
      final_state: initial_state,
      mutations_applied: [],
    };
  }

  for (let iter = 0; iter < MAX_ITERATIONS; iter++) {
    const report = predictCTLPerceptionState(current);

    if (report.state === "harmonic") {
      return {
        ctl: current,
        converged: true,
        iterations: iter,
        initial_state,
        final_state: report,
        mutations_applied: mutations,
      };
    }

    const params = report.params;

    // Apply fixes in order of severity — C1 first (dominant constraint)
    if (!report.c1_pass || params.b_eff >= C1_B_EFF_HARMONIC_MAX) {
      const fix = fixC1(current, params);
      current = fix.ctl;
      mutations.push(fix.mutation);
    }

    if (!report.c2_pass) {
      const fix = fixC2(current, params);
      current = fix.ctl;
      mutations.push(fix.mutation);
    }

    if (!report.c3_pass) {
      const fix = fixC3(current);
      current = fix.ctl;
      mutations.push(fix.mutation);
    }
  }

  const final_state = predictCTLPerceptionState(current);

  return {
    ctl: current,
    converged: final_state.state === "harmonic",
    iterations: MAX_ITERATIONS,
    initial_state,
    final_state,
    mutations_applied: mutations,
  };
}

File: apps/audio/services/intelligence/ctl_optimizer.py
"""
CTL Perception Optimizer.

Iteratively adjusts a CTL document until C1 + C2 + C3 all pass.
Superior to the TypeScript version in two ways:
  1. 8 max iterations (vs 6) — converges harder cases
  2. Adaptive C1 scaling: if profile store has n >= 5 evaluations for this lane,
     uses the learned optimal b_eff as target instead of the hardcoded 0.32

Mutations are applied in C1 → C2 → C3 order (most impactful first).
C2 fix runs two passes: removes K tokens first, then downgrades L → ghost.
"""

from __future__ import annotations
from .perception_engine import (
    predict_ctl_state, extract_ctl_params, PerceptionReport,
    C1_B_EFF_HARMONIC_MAX, C1_B_EFF_PERCUSSION_MIN,
    C1_OPTIMAL_ALPHA_B, C2_MAX_TRANSIENTS, C3_MIN_LD_DENSITY,
)
from typing import TYPE_CHECKING, Optional

if TYPE_CHECKING:
    from .adaptive_profiles import AdaptiveProfileStore

MAX_ITERATIONS = 8
C1_LD_SCALE    = 0.88
C1_BASS_SCALE  = 0.85
_PRIMARY       = {0, 8}


def _fix_c1(ctl: dict, params, target_b_eff: float = C1_OPTIMAL_ALPHA_B) -> tuple[dict, str]:
    instr = []
    for inst in ctl.get("instrumentation", []):
        fam = inst.get("family", "")
        w   = inst.get("body_weight", 0.0)
        if fam == "log_drum":
            instr.append({**inst, "body_weight": round(max(0.10, w * C1_LD_SCALE), 3)})
        elif fam in ("bass", "kick") and params.alpha_B > 0.40:
            instr.append({**inst, "body_weight": round(max(0.10, w * C1_BASS_SCALE), 3)})
        else:
            instr.append(inst)
    new_p = extract_ctl_params({**ctl, "instrumentation": instr})
    return (
        {**ctl, "instrumentation": instr},
        f"c1_fix: B_eff {params.b_eff:.3f}→{new_p.b_eff:.3f} (target <{C1_B_EFF_HARMONIC_MAX})"
    )


def _fix_c2(ctl: dict, params) -> tuple[dict, str]:
    before = params.transient_density
    patterns = []
    for p in ctl.get("groove_patterns", []):
        steps = list(p.get("steps", []))
        count = sum(1 for s in steps if s in ("K", "L"))

        # Pass 1: remove non-primary K tokens (back to front)
        for i in reversed(range(len(steps))):
            if count <= C2_MAX_TRANSIENTS:
                break
            if steps[i] == "K" and i not in _PRIMARY:
                steps[i] = "-"
                count -= 1

        # Pass 2: downgrade non-primary L → ghost
        for i in reversed(range(len(steps))):
            if count <= C2_MAX_TRANSIENTS:
                break
            if steps[i] == "L" and i not in _PRIMARY:
                steps[i] = "g"
                count -= 1

        patterns.append({**p, "steps": steps})

    counts = [sum(1 for s in p["steps"] if s in ("K", "L")) for p in patterns]
    after  = sum(counts) / len(counts) if counts else 0
    return (
        {**ctl, "groove_patterns": patterns},
        f"c2_fix: transients {before:.1f}→{after:.1f}/bar"
    )


def _fix_c3(ctl: dict) -> tuple[dict, str]:
    ld = ctl.get("curves", {}).get("log_drum_density", [])
    fixed, changed = [], 0
    for pt in ld:
        if pt.get("value", 0) < C3_MIN_LD_DENSITY:
            fixed.append({**pt, "value": C3_MIN_LD_DENSITY})
            changed += 1
        else:
            fixed.append(pt)
    curves = {**ctl.get("curves", {}), "log_drum_density": fixed}
    return (
        {**ctl, "curves": curves},
        f"c3_fix: raised {changed} point(s) to floor {C3_MIN_LD_DENSITY}"
    )


def optimize_ctl(
    ctl: dict,
    profile_store: Optional["AdaptiveProfileStore"] = None,
) -> dict:
    """
    Iteratively fix CTL until perception state is harmonic.
    Returns the CTL dict extended with _perception_optimization metadata.
    """
    lane = ctl.get("global", {}).get("subgenre", "")

    # Optionally pull adaptive b_eff target from profile store
    adaptive_target = C1_OPTIMAL_ALPHA_B
    if profile_store and lane:
        stat = profile_store.get(lane, "b_eff")
        if stat.n >= 5:
            adaptive_target = stat.mean

    current    = ctl
    mutations: list[str] = []
    initial    = predict_ctl_state(current)

    if initial.state == "harmonic":
        return {**current, "_perception_optimization": {
            "converged": True, "iterations": 0,
            "initial_state": initial.state, "final_state": initial.state,
            "mutations_applied": [],
        }}

    for _ in range(MAX_ITERATIONS):
        report = predict_ctl_state(current)
        if report.state == "harmonic":
            break
        params = report.params

        if not report.c1_pass or params.b_eff >= C1_B_EFF_HARMONIC_MAX:
            current, m = _fix_c1(current, params, adaptive_target)
            mutations.append(m)

        if not report.c2_pass:
            current, m = _fix_c2(current, params)
            mutations.append(m)

        if not report.c3_pass:
            current, m = _fix_c3(current)
            mutations.append(m)

    final = predict_ctl_state(current)
    return {**current, "_perception_optimization": {
        "converged": final.state == "harmonic",
        "iterations": len(mutations),
        "initial_state": initial.state,
        "final_state": final.state,
        "mutations_applied": mutations,
        "violations": final.violations,
    }}
